### PR TITLE
add openNav, closeNav, openSubnav, closeSubnav events

### PR DESCRIPTION
**Pull request by JBCSU**
# READY FOR REVIEW

# Summary
In Scott's [event calendar project](http://events-dev.stanford.edu/), the **By Date** subnav contains a date picker. For accessibility reasons, he needs to place the focus on the date picker element when that subnav is opened. The cleanest way for him to do that is to add an event listener that responds whenever the appropriate subnav opens. He therefore needs an event to be triggered when a subnav opens. It seemed like it might also be useful to trigger events when a subnav closes, and when a mobile nav opens or closes.

# Needed By (Date)
Soon-ish, so Scott can respond to his accessibility issue.

# Urgency
Meh.

# Steps to Test

1. Pull the branch
1. `npm run build`
1. `npm run styleguide`
1. Visit your local dev instance in your browser and scroll down to the [Main Navigation](http://decanter.local/section-components.html#kssref-components-main-navigation) component
1. Open your console and paste the following commands
```
navs = document.querySelectorAll('.su-main-nav');
navs[1].addEventListener('openNav', (ev) => {console.log('mobile nav 1 OPENED');});
navs[1].addEventListener('closeNav', (ev) => {console.log('mobile nav 1 CLOSED');});
subnavs = document.querySelectorAll('.su-main-nav__item--parent');
subnavs[4].addEventListener('openSubnav', (ev) => {console.log('subnav 4 OPENED');});
subnavs[4].addEventListener('closeSubnav', (ev) => {console.log('subnav 4 CLOSED');});
subnavs[5].addEventListener('openSubnav', (ev) => {console.log('subnav 5 OPENED');});
subnavs[5].addEventListener('closeSubnav', (ev) => {console.log('subnav 5 CLOSED');});
```
6. With your console still open, try various ways of opening and closing  the **Useful Websites** and **Tools & Documentation** subnavs in the first example nav (**Default styling**). Notice the messages in the console.
6. Make your browser small enough so the example navs switch to mobile. With your console still open, try various ways of opening and closing  the first example nav (**Default styling**). Notice the messages in the console.

# Affected Projects or Products
None. However, we need to figure out how Scott can get these changes. Right now he's fetching the `5.0.0` tag in the decanter repo to keep his environment stable. He probably shouldn't be pointing directly to `master`, as that's still volatile. Should we create a `5.0.1` tag? Or should we keep this branch alive after it's merged and have him point to this branch?

# Associated Issues and/or People
- @stocker

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

**Comment by josephgknox** [comment]
@JBCSU are any of the CC issues ones that you think we should correct before merging? Does out linter need to be updated to ignore such flags (looks like 7 of the 8 issues are the same)?

**Comment by JBCSU** [comment]
> @JBCSU are any of the CC issues ones that you think we should correct before merging? Does out linter need to be updated to ignore such flags (looks like 7 of the 8 issues are the same)?

7 of them are effectively complaining that I commented my code: the complaint is that the total line length, including my comments, is >80 characters. IMO we should change our linter to either not count comments in character counts, or allow longer lines. Or, IMO, both. 80 characters was the length of a punched card. It's 2019 and we're all using IDE's. I think we could go to 100, 120, maybe longer? And I think comment characters should be ignored, or perhaps even subtracted from overall line length! Comments are Good Things.

The other one refers to the "cognitive complexity" I introduced by only collapsing subnavs if they're actually open. Previously I "collapsed" them unconditionally because it was just DOM tweaking, which is very cheap. But now that I'm actually firing an event, which is much more costly, I only want to fire the event if the subnav actually collapsed. That's a 14 line function that IMO is eminently readable. I am 100% confident that anyone looking at that function could understand it perfectly well. So no, I don't think that CC complaint needs to be addressed either.

But those are my opinions. You guys are my reviewers, so you tell me.

**Comment by josephgknox** [comment]
As always, thanks so much for the detailed reply! This makes sense to me, and I am behind each of your opinions here. My vote would be to update our linters to allow for longer lines as well as not count comments. Let's see if others have thoughts, too.

File: core/src/js/utilities/events.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/**
+ * Create an event with the specified name in a browser-agnostic way.
+ * @param {string} eventName - the name of the event
```
Comment (sherakama) [inline core/src/js/utilities/events.js]: This parameter is not {string} bound but rather mixed as it accepts strings or functions.

File: core/src/js/utilities/events.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ * @return {Event} - instance of event which can be dispatched / listened for
+ */
+export const createEvent = (eventName = '') => {
```
Comment (sherakama) [inline core/src/js/utilities/events.js]: Do you need the default string? I'd vote no as it is not useful to have an empty string passed in.

File: core/src/js/utilities/events.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  if (typeof eventName !== 'string' || eventName.length <= 0) {
+    return null;
```
Comment (sherakama) [inline core/src/js/utilities/events.js]: Do we want this to quietly fail or do we want to throw an exception here?

File: core/src/js/components/main-nav/Nav.js [hunk 4]
```diff
@@ -47,6 +48,10 @@ export default class Nav {
       }
     );
 
+    // add custom events to alert others when the mobile nav opens or closes
```
Comment (sherakama) [inline core/src/js/components/main-nav/Nav.js]:
```suggestion
    // Add custom events to alert others when the mobile nav opens or closes.
```

Comment (sherakama) [reply core/src/js/components/main-nav/Nav.js]: Can we move this new block up to line 35/36 so that the instance property assignments are all together before the logic statements?

File: core/src/js/components/main-nav/Nav.js [hunk 5]
```diff
@@ -265,15 +270,21 @@ export default class Nav {
     if (focusOnFirst) {
       this.focusOn('first'); // Focus on the first top level link
     }
+    // alert others the mobile nav has opened
```
Comment (sherakama) [inline core/src/js/components/main-nav/Nav.js]:
```suggestion
    // Alert others the mobile nav has opened.
```

File: core/src/js/components/main-nav/Nav.js [hunk 6]
```diff
@@ -265,15 +270,21 @@ export default class Nav {
...
   }
 
   /**
    * Mark this mobile closed, and restore the button text to what it was
    * initially.
    */
   closeMobileNav() {
-    this.setExpanded('false');
-    this.toggle.innerText = this.toggleText;
+    if (this.isExpanded()) {
```
Comment (sherakama) [inline core/src/js/components/main-nav/Nav.js]: Why do we need to check for expanded here? Does it matter?

File: core/src/js/components/main-nav/Nav.js [hunk 7]
```diff
@@ -265,15 +270,21 @@ export default class Nav {
...
+      this.setExpanded('false');
+      this.toggle.innerText = this.toggleText;
+      // alert others the mobile nav has closed
```
Comment (sherakama) [inline core/src/js/components/main-nav/Nav.js]:
```suggestion
      // Alert others the mobile nav has closed.
```

File: core/src/js/components/main-nav/NavItem.js [hunk 8]
```diff
@@ -35,6 +35,10 @@ export default class NavItem {
       // Maintain global list of subnavs for closeAllSubNavs().
       theSubNavs.push(this);
       this.item.addEventListener('click', this);
+
+      // add custom events to alert others when a subnav opens or closes
```
Comment (sherakama) [inline core/src/js/components/main-nav/NavItem.js]:
```suggestion
      // Add custom events to alert others when a subnav opens or closes.
```

File: core/src/js/components/main-nav/NavItem.js [hunk 9]
```diff
@@ -140,10 +145,13 @@ export default class NavItem {
    */
   closeSubNav(focusOnTrigger = false) {
     if (this.isSubNavTrigger()) {
-      this.item.classList.remove('su-main-nav__item--expanded');
-      this.setExpanded('false');
-      if (focusOnTrigger) {
-        this.link.focus();
+      if (this.isExpanded()) {
```
Comment (sherakama) [inline core/src/js/components/main-nav/NavItem.js]: Why do we need to check if the item is expanded here? Is it destructive to execute the close items if it wasn't open?

File: core/src/js/components/main-nav/main-nav.js [hunk 10]
```diff
@@ -6,22 +6,23 @@ document.addEventListener('DOMContentLoaded', event => {
...
   const navs = document.querySelectorAll('.' + navClass);
 
-  // ---------------------------------------------------------------------------
-  // Execution Code.
-  // ---------------------------------------------------------------------------
-
   // Process each nav.
   let firstZindex;
   navs.forEach((nav, index) => {
+    // remove the class that formats the nav for browsers with javascript disabled
```
Comment (sherakama) [inline core/src/js/components/main-nav/main-nav.js]:
```suggestion
    // Remove the class that formats the nav for browsers with javascript disabled.
```

File: core/src/js/components/main-nav/main-nav.js [hunk 11]
```diff
@@ -6,22 +6,23 @@ document.addEventListener('DOMContentLoaded', event => {
...
+    nav.classList.remove('no-js');
```
Comment (sherakama) [inline core/src/js/components/main-nav/main-nav.js]: This is happening in the template file. Can we remove that javascript code from there now?

File: core/src/js/components/main-nav/main-nav.js [hunk 12]
```diff
@@ -6,22 +6,23 @@ document.addEventListener('DOMContentLoaded', event => {
...
+
+    // create an instance of Nav, which in turn creates appropriate instances of NavItem
```
Comment (sherakama) [inline core/src/js/components/main-nav/main-nav.js]:
```suggestion
    // Create an instance of Nav, which in turn creates appropriate instances of NavItem.
```

File: core/src/js/components/main-nav/main-nav.js [hunk 13]
```diff
@@ -6,22 +6,23 @@ document.addEventListener('DOMContentLoaded', event => {
...
     const theNav = new Nav(nav);
+
+    // remember the nav for closeAllMobileNavs()
```
Comment (sherakama) [inline core/src/js/components/main-nav/main-nav.js]:
```suggestion
    // Remember the nav for closeAllMobileNavs().
```

File: core/src/js/components/main-nav/main-nav.js [hunk 14]
```diff
@@ -6,22 +6,23 @@ document.addEventListener('DOMContentLoaded', event => {
...
     theNavs.push(theNav);
 
+    // manage zindexes in case there are multiple navs near enough for subnavs to overlap
```
Comment (sherakama) [inline core/src/js/components/main-nav/main-nav.js]:
```suggestion
    // Manage zindexes in case there are multiple navs near enough for subnavs to overlap
```

File: core/src/js/components/main-nav/main-nav.js [hunk 15]
```diff
@@ -6,22 +6,23 @@ document.addEventListener('DOMContentLoaded', event => {
...
+    // extremely unlikely, but it happens in the styleguide
```
Comment (sherakama) [inline core/src/js/components/main-nav/main-nav.js]:
```suggestion
    // extremely unlikely, but it happens in the styleguide.
```

File: core/src/js/utilities/events.js [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+  }
+  if (typeof Event == 'function') { // modern browsers
```
Comment (sherakama) [inline core/src/js/utilities/events.js]:
```suggestion
  // Modern browsers.
  if (typeof Event == 'function') {
```

File: core/src/js/utilities/events.js [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+    return new Event(eventName);
+  }
+  else { // IE :(
```
Comment (sherakama) [inline core/src/js/utilities/events.js]:
```suggestion
  // IE :(
  else {
```

**Comment by sherakama** [comment]
@JBCSU the 80 char limit is pretty common across many projects. Keeping the comments above the item in which they refer to and at 80 characters keep formatting consistent. 

https://nickjanetakis.com/blog/80-characters-per-line-is-a-standard-worth-sticking-to-even-today
https://hackernoon.com/does-column-width-of-80-make-sense-in-2018-50c161fbdcf6

Seems like there is some movement to increase this line limit to upwards of 120. 
Looks like airbnb's standards are at 100: https://github.com/airbnb/javascript/pull/639/files
Bootstrap disabled theirs.

I think we should keep a max character width in our code and I am open to increasing it. Should it be 100 or 120?

File: core/src/js/utilities/events.js (hunk 1 above)
Comment (JBCSU) [reply core/src/js/utilities/events.js]: According to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Event/Event), the first parameter to `new Event()` is a DOMString, and "[DOMString is mapped directly to a String](https://developer.mozilla.org/en-US/docs/Web/API/DOMString)." Also, my wrapper function explicitly test for string, and fails if it's not a (non-empty) string.

File: core/src/js/utilities/events.js (hunk 3 above)
Comment (JBCSU) [reply core/src/js/utilities/events.js]: Normally I like to defend against errors, but in this case I guess a compile time error is better than a runtime error. I'll remove the default.

File: core/src/js/components/main-nav/Nav.js (hunk 4 above)
Comment (JBCSU) [reply core/src/js/components/main-nav/Nav.js]: The only logic that's above it is initializing `this.items`. I like to keep the initializing of properties next to the declarations. That way I only have to look in 1 place when I need to know what the property is.

File: core/src/js/components/main-nav/Nav.js (hunk 5 above)
Comment (JBCSU) [reply core/src/js/components/main-nav/Nav.js]: I really don't like the sentence-like construction of comments. Can we agree to disagree on that comment style?

File: core/src/js/components/main-nav/Nav.js (hunk 6 above)
Comment (JBCSU) [reply core/src/js/components/main-nav/Nav.js]: I didn't have it before. But now that we're firing an event when the nav closes, yes, I think it matters:

- Firing an event is much more expensive than DOM tweaking. I don't think we should incur that cost unless we need to.
- If someone is listening for the `closeNav` event, I think they'll only want to be notified if the nav actually closed. Right now we call `closeMobileNav()` anytime anyone clicks outside of a nav. We don't want to fire `closeNav` every time that happens.

File: core/src/js/components/main-nav/NavItem.js (hunk 9 above)
Comment (JBCSU) [reply core/src/js/components/main-nav/NavItem.js]: See comment above about `closeMobileNav()`. Same reasoning applies here.

File: core/src/js/components/main-nav/main-nav.js (hunk 11 above)
Comment (JBCSU) [reply core/src/js/components/main-nav/main-nav.js]: I thought I still saw a brief FOUC when I removed it from the template file, but it seems ok now. I'm happy to remove it.

File: core/src/js/components/main-nav/main-nav.js (hunk 15 above)
Comment (JBCSU) [reply core/src/js/components/main-nav/main-nav.js]: If I were totally unfamiliar with Decanter's nav, but reading through the code in anticipation of using it in a project, I would be deeply puzzled by that zIndex loop. I would be reassured by knowing that the problem it's trying to solve is in fact extremely unlikely in production.

File: core/src/js/utilities/events.js (hunk 16 above)
Comment (JBCSU) [reply core/src/js/utilities/events.js]: I think there's room for both styles of commenting. I like comments on their own line when they're introducing a distinct chunk of code. That style of commenting helps me chunk up code mentally, in part because it interrupts my reading of the code. I like comments at the end of the line when they're just asides to help clarify a particular line or two of code. They're off to the side, so they don't interrupt my reading of the code.

**Comment by JBCSU** [comment]
> I think we should keep a max character width in our code and I am open to increasing it. Should it be 100 or 120?

If trailing comments count, then I vote for 120. If only code counts, then 100 seems fine.

File: core/src/js/components/main-nav/Nav.js (hunk 4 above)
Comment (sherakama) [reply core/src/js/components/main-nav/Nav.js]: In javascript variables are hoisted. See: https://www.w3schools.com/js/js_hoisting.asp

It is best practices to always declare your variables and properties at the top. 

> To avoid bugs, always declare all variables at the beginning of every scope.
> Since this is how JavaScript interprets the code, it is always a good rule.

```
    this.elem = elem;
    this.topNav = this.getTopNav();
    
    // Add custom events to alert others when the mobile nav opens or closes
    this.openEvent = createEvent('openNav'); // dispatched in this.openMobileNav()
    this.closeEvent = createEvent('closeNav'); // dispatched in this.closeMobileNav()

    this.toggle = elem.querySelector(elem.tagName + ' > button');
    this.toggleText = this.toggle ? this.toggle.innerText : '';
    this.items = [];

    // If this is a subnav, we need the correpsonding HTMLElement for
    // .querySelector()
    if (elem instanceof NavItem) {
      elem = elem.item;
    }
   
    let items = elem.querySelectorAll(elem.tagName + ' > ul > li');
    items.forEach(
      item => {
        this.items.push(new NavItem(item, this));
      }
    );
```

File: core/src/js/components/main-nav/Nav.js (hunk 5 above)
Comment (sherakama) [reply core/src/js/components/main-nav/Nav.js]: This is a convention that has been hammered into me by multiple other projects. I don't see anything that would make this a requirement in this project. Checking a few of the most popular projects on npm.org, I see the following:

lodash: Uses sentence structure
express: Doesn't use sentence structure
react: Capital start but no period.
chalk: Mixed, no format.
request: Capital start but no period.
commander: Capital start but no period.
Moment: Uses sentence structure
debug: Uses capitals but no period.

How about we split the difference? Comments must start with a capital letter but no punctuation is required at the end? That seems to be the common pattern.

File: core/src/js/components/main-nav/main-nav.js (hunk 15 above)
Comment (sherakama) [reply core/src/js/components/main-nav/main-nav.js]: If you feel strongly about this comment then I am fine to leave it along. My experience in working with open source projects is that I shouldn't try to guess what crazy scenarios another developer may have. Forces unknown can lead to wacky scenarios.

File: core/src/js/utilities/events.js (hunk 16 above)
Comment (sherakama) [reply core/src/js/utilities/events.js]: Checking the top packages on npmjs.org

lodash: Seems to be above but mostly just has function comments
request: mixed, some above and some inline
chalk: mixed, some above and some inline
react: above
debug: above
moment: mixed, some above and some inline
express: mixed, commonly inline.

Checking wordpress standards it appears they use a modified jQuery convention:
- https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/
- https://contribute.jquery.org/style-guide/js/#comments

> Comments are always preceded by a blank line. Comments start with a capital first letter, but don't require a period at the end, unless you're writing full sentences. There must be a single space between the comment token and the comment text.

I try to stick to the conventions of the project I intend to be working in as it keeps everything similar but it looks like it is fairly mixed in some of the top npm packages.

File: core/src/js/components/main-nav/Nav.js (hunk 6 above)
Comment (sherakama) [reply core/src/js/components/main-nav/Nav.js]: Sounds like a good idea to me. Out of curiosity, what did you use to measure the expense of the event call?

File: core/src/js/components/main-nav/NavItem.js (hunk 9 above)
Comment (sherakama) [reply core/src/js/components/main-nav/NavItem.js]: Thanks, GTG here.

**Comment by sherakama** [comment]
Ok, 120 it is. 

Just as an FYI, it looks like Wordpress uses the 80 character limit. https://make.wordpress.org/core/handbook/best-practices/coding-standards/css/#commenting

File: core/src/js/utilities/events.js (hunk 1 above)
Comment (sherakama) [reply core/src/js/utilities/events.js]: Oh, I totally read the if statements wrong. I thought the second check where the if statement is checking to see if Event is a function was on the input parameter. You are 100% right. Apologies for this comment.

File: core/src/js/components/main-nav/Nav.js (hunk 6 above)
Comment (JBCSU) [reply core/src/js/components/main-nav/Nav.js]: > Out of curiosity, what did you use to measure the expense of the event call?

Intuition. Plus, I watched my event handlers getting called **WAY** more often than they needed to be. :)

File: core/src/js/utilities/events.js (hunk 16 above)
Comment (JBCSU) [reply core/src/js/utilities/events.js]: > I try to stick to the conventions of the project I intend to be working in as it keeps everything similar but it looks like it is fairly mixed in some of the top npm packages.

I do try to stick to the conventions of the project I’m working in. But since this is the very first Javascript in the project, I thought there might be some flexibility around what those conventions are.

Also, to my mind, conventions are not ends in themselves. They’re there to make reading and understanding the code easier. I honestly cannot imagine a scenario where a comment beginning with a lower case letter vs. an upper case letter would affect anyone’s ability to read and understand the code. I therefore don’t think it’s in anyone’s best interests to bother having a convention about that. I accept that some people do make conventions about that, but to my mind that’s not a productive use of anyone’s time.

Of course discussing comment style at this length is also not a productive use of our time. I’ve changed the code per your requests.

File: core/src/js/components/main-nav/Nav.js (hunk 4 above)
Comment (JBCSU) [reply core/src/js/components/main-nav/Nav.js]: I updated the constructor per your request.

File: core/src/js/components/main-nav/main-nav.js (hunk 15 above)
Comment (JBCSU) [reply core/src/js/components/main-nav/main-nav.js]: I modified the comment to say "Rare, but..."

File: core/src/js/utilities/events.js (hunk 16 above)
Comment (sherakama) [reply core/src/js/utilities/events.js]: I'd argue that the very first of something is the best time to start a convention. Instead of letting something snowball into work later, nip it early and carry on. Starting early is also a benefit to the early contributors as they get the biggest say into what the convention should be.

I agree that capitalization and punctuation of comments is not a make or break scenario for readability but rather, my concerns stem from having consistency in style. Consistency plays a big part into readability and is good for the overall health of the project. Many other people much smarter than myself have had long discussions on the merits of each style and have created unified standards for their projects. I'd like to stay in line with the projects we intend to use this code in and when there isn't one, pick from other popular projects. 

My goal of these reviews is not to have long pedantic conversations about style, but rather, to call out and pick a standard when I see a difference. Please bear with this process as it gets quicker as we go.

File: core/src/templates/components/main-nav/main-nav.twig [hunk 18]
```diff
@@ -65,8 +65,3 @@
     %}
   {% endif %}
 </nav>
-<script>
```
Comment (sherakama) [inline core/src/templates/components/main-nav/main-nav.twig]: woo hoo. Thanks for this!.

File: core/src/js/utilities/events.js (hunk 16 above)
Comment (JBCSU) [reply core/src/js/utilities/events.js]: I trust that we both have the best interests of the project at heart. But I think the comment thread of what was meant to be a minor PR is likely not the best forum for proposing, discussing or determining these kinds of coding standards, especially when there's clearly some disagreement. If we want to invest time in coming to a shared consensus on standards, I'm happy to discuss it in a call, either 1-1 or in a regular meeting. Alternatively, I did propose including a coding standards issue in our upcoming Decanterpalooza, for just this reason. But in the meantime, for the sake of this PR, I believe I have made all the changes you requested. I'm happy to shake hands, reaffirm our shared commitment to writing excellent code that benefits Stanford, and go forth and conquer more components! :)

File: core/src/templates/components/main-nav/main-nav.twig (hunk 18 above)
Comment (JBCSU) [reply core/src/templates/components/main-nav/main-nav.twig]: Fer shur. That needed to go!

**Comment by sherakama** [comment]
Thank you for the learning opportunities and valuable feedback.

File: core/src/js/utilities/events.js (hunk 16 above)
Comment (sherakama) [reply core/src/js/utilities/events.js]: That's a really good point. We could speed things up by spending some time hashing out the lint files. +1 to getting to those tickets sooner than later.
